consensus/propagation/have_wants.go

```diff
@@ -63,58 +64,133 @@ func (blockProp *Reactor) handleHaves(peer p2p.ID, haves *proptypes.HaveParts) {
 		return
 	}
 
-	reqLimit := 1
-
-	// if enough requests have been made for the parts, don't request them.
-	for _, partIndex := range hc.GetTrueIndices() {
-		reqs := blockProp.countRequests(height, round, partIndex)
-		if len(reqs) >= reqLimit {
-			hc.SetIndex(partIndex, false)
-			// mark the part as fully requested.
-			fullReqs.SetIndex(partIndex, true)
-		}
-		// don't request the part from this peer if we've already requested it
-		// from them.
-		for _, p := range reqs {
-			// p == peer means we have already requested the part from this peer.
-			if p == peer {
-				hc.SetIndex(partIndex, false)
-			}
+	for _, index := range hc.GetTrueIndices() {
+		select {
+		case <-blockProp.ctx.Done():
+		case p.requestChan <- request{height: height, round: round, index: uint32(index)}:
 		}
 	}
 
-	if hc.IsEmpty() {
-		return
+	blockProp.broadcastHaves(haves, peer, int(parts.Total()))
+}
+
+// perPeerPerBlockConcurrentRequestLimit returns the maximum number of requests for a given block by peer.
+// it helps avoid downloading the whole block from a single peer.
+func perPeerPerBlockConcurrentRequestLimit() int64 {
+	// TODO implement
+	return 300
+}
+
+func (blockProp *Reactor) wantsSendingRoutine(ps *PeerState) {
+	var requestsBA *bits.BitArray
+	height := int64(0)
+	round := int32(0)
+	batchRequestCount := int64(0)
+	for {
+		select {
+		case <-blockProp.ctx.Done():
+			return
+		case req, ok := <-ps.requestChan:
+			if !ok {
+				return
+			}
+
+			// concurrent per-peer per-block request limit reached
+			if ps.requestCount.Load()+batchRequestCount >= perPeerPerBlockConcurrentRequestLimit() {
+				// we reached the limit of requests, we request what we can
+				if requestsBA != nil && !requestsBA.IsEmpty() {
+					blockProp.sendWant(ps, height, round, requestsBA)
+					requestsBA = nil
+					batchRequestCount = 0
+				}
+				// wait for a part to be received before continuing
+				select {
+				case <-blockProp.ctx.Done():
+					return
+				case _, ok = <-ps.receivedPart:
+					if !ok {
+						return
+					}
+				}
+			}
+
+			// if we're at a new height, we can drop the previous requests
+			if height != req.height || (height == req.height && round != req.round) {
+				requestsBA = nil
+				batchRequestCount = 0
+				height = req.height
+				round = req.round
+			}
```

Comment:
I still think this is unecessarily bug prone and laggy when we could simply look up the relevant height instead of relying on the requests